### PR TITLE
Read email from SSO profile – WEB-437

### PR DESCRIPTION
**Pull request by AlexTugarev**
## Description
This PR introduces the `Identity.lastSigninTime` attribute, which is updated as `User.identities[signed in idp].email` on each sign-in into an Org-owned account. 

The email address provided by the IdP is to be rendered for Org-owned accounts. If the account is Org-owned we'll take the most recent email address, which takes into account that the Org might have updated the SSO config and therefore an older identity remains in the database. 

## Related Issue(s)

Fixes WEB-437

## How to test
* Sign-in into the Org https://at-sso-email.preview.gitpod-dev.com/login/acme
* Configure a connection to a Git provider with a different email address
   * This is blocked by WEB-444 as I'm not able to connect from within a simple SSO account
* See the email address is unchanged on /user/settings

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - at-sso-email</li>
	<li><b>🔗 URL</b> - <a href="https://at-sso-email.preview.gitpod-dev.com/workspaces" target="_blank">at-sso-email.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [x] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>
 
/hold

### CODE REVIEW

**Comment by AlexTugarev** [comment]
🙈 first attempt didn't work well. 

```
POST https://at-sso-email.preview.gitpod-dev.com/public-api/gitpod.experimental.v1.OIDCService/ListClientConfigs 403
```
```
ConnectError: [permission_denied] user 580a6b50-de4c-4d4a-8c0d-b02da77a67f2 is not owner of organization 278ed4c8-385e-4a2a-af8b-7817e72de1da
```

The alternative approach would be to mark SSO identities as such.

**Comment by AlexTugarev** [comment]
Screenshot 2023-06-05 at 16 24 40

Now the `lastSigninTime` seems to be updated properly on each sign-in. Having that in place we can assume that the recently marked identity of an Org-owned account can be the source of the email address provided by the IdP.

**Comment by easyCZ** [comment]
From the PR, I don't really understand how this fixes [WEB-437](https://linear.app/gitpod/issue/WEB-437) and why we'd need to store the last time that record was accessed. It feels odd to touch a record every time you read it.

Would you be able to explain to me how storing the last sign-in time solves the issue of a different email being considered the primary email?

Could we instead solve it by allowing the user to pick the primary email, and in the case of org owned accounts, to not allow choosing but default to the single identity?

**Comment by AlexTugarev** [comment]
@easyCZ, happy to discuss this. The preceded discussion took place in [here](https://linear.app/gitpod/issue/WEB-437/primary-github-email-overwrites-sso-email-in-case-of-conflict#comment-981f7cd8). 

The goal is to read the email address from `User.identities[signed in idp].email` if the account is Org-owned. We've no other options to track the sign-ins, and storing the `lastSigninTime` seems to be least invasive method of tracking this. 

> Would you be able to explain to me how storing the last sign-in time solves the issue of a different email being considered the primary email?

We'd store the `lastSigninTime` on sign-ins only, not on every access, which would be bogus. Given that, for Org-owned accounts the most recent record is the one containing the most recent email address. And users of Org-owned accounts are not able to change this value. 

This is different for regular (non Org-owned) account, and I also didn't want to touch semantics in there. We are using the `additionalData` bucked to store the user provided value.

File: components/gitpod-db/src/typeorm/migration/1685955742395-AddLastSigninTimeToIdentity.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import { columnExists } from "./helper/helper";
+
+const table = "d_b_identity";
+const column = "lastSigninTime";
+
+export class AddLastSigninTimeToIdentity1685955742395 implements MigrationInterface {
+    public async up(queryRunner: QueryRunner): Promise<void> {
+        if (!(await columnExists(queryRunner, table, column))) {
+            await queryRunner.query(
+                `ALTER TABLE ${table} ADD COLUMN ${column} varchar(30) NOT NULL DEFAULT '', ALGORITHM=INPLACE, LOCK=NONE`,
```
Comment (easyCZ) [inline components/gitpod-db/src/typeorm/migration/1685955742395-AddLastSigninTimeToIdentity.ts:17]: What if we used an actual DATETIME type here?

File: components/gitpod-db/src/typeorm/entity/db-identity.ts [hunk 2]
```diff
@@ -36,6 +36,15 @@ export class DBIdentity implements Identity {
     @Column()
     deleted?: boolean;
 
+    @Column({
+        default: "",
+        transformer: Transformer.MAP_EMPTY_STR_TO_UNDEFINED,
+    })
+    lastSigninTime?: string;
```
Comment (easyCZ) [inline components/gitpod-db/src/typeorm/entity/db-identity.ts:43]: Would be nice to send the list of Identities ordered by the `lastSigninTime` from the backend, but I suspect that could be harder to given the one-to-many relationship

**Review by easyCZ — APPROVED** [review]
LGTM

File: components/gitpod-protocol/src/protocol.ts [hunk 3]
```diff
@@ -657,7 +645,10 @@ export interface Identity {
     primaryEmail?: string;
     /** This is a flag that triggers the HARD DELETION of this entity */
     deleted?: boolean;
-    // readonly identities cannot be modified by the user
+    // The last time this entry was touched during a signin.
+    lastSigninTime?: string;
+
+    // @deprecated as no longer in use since '19
     readonly?: boolean;
```
[the remark this answers is not in the document]
Comment (AlexTugarev) [reply components/gitpod-protocol/src/protocol.ts:652]: absolutely, didn't want to do in here.

File: components/gitpod-db/src/typeorm/entity/db-identity.ts (hunk 2 above)
Comment (AlexTugarev) [reply components/gitpod-db/src/typeorm/entity/db-identity.ts:43]: that's actually a good idea.

**Pull request merged**